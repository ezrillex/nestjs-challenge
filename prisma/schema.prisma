// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
// Try Prisma Accelerate: https://pris.ly/cli/accelerate-init

// npx prisma migrate dev --name the_name
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Testing {
  id   String @id @default(uuid())
  test String
}

enum roles {
  customer
  manager
  admin
  anonymous // public
}

model Users {
  id                                 String     @id @default(uuid())
  first_name                         String     @db.VarChar(60)
  last_name                          String     @db.VarChar(120)
  email                              String     @unique
  password                           String     @db.Char(60) // hashed and salted password
  created_at                         DateTime   @default(now())
  role                               roles      @default(customer)
  failed_login_attempts              Int        @default(0)
  failed_login_attempts_timestamps   DateTime[] @default([])
  last_login_at                      DateTime?
  password_reset_requests            Int        @default(0)
  password_reset_requests_timestamps DateTime[] @default([])
  password_reset_token               String?
  password_last_updated              DateTime   @default(now())
}
