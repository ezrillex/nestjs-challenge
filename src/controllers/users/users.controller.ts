import { Controller, Get, HttpCode, HttpStatus } from '@nestjs/common';

@Controller('users')
export class UsersController {
  @Get()
  @HttpCode(HttpStatus.OK)
  async getUsers() {
    return 'Hello World!';
  }
}
