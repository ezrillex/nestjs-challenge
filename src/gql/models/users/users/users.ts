import { Field, ID, ObjectType } from '@nestjs/graphql';
import { LikesOfProducts } from '../../products/likes/likes_of_products/likes_of_products';
import { IsEmail } from 'class-validator';

// partial implementation of actual table due to graphql not dealing with auth and roles
@ObjectType()
export class Users {
  @Field(() => ID)
  id: string;

  @Field(() => String)
  first_name: string;

  @Field(() => String)
  last_name: string;

  @IsEmail()
  @Field(() => String)
  email: string;

  @Field(() => [LikesOfProducts])
  likes_products: LikesOfProducts[];
}
