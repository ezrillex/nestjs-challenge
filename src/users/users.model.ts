import { Field, ID, ObjectType } from '@nestjs/graphql';
import { IsEmail } from 'class-validator';
import { LikesOfProducts } from '../likes/LikesOfProducts.model';
import { roles } from '@prisma/client';

@ObjectType()
export class Users {
  @Field(() => ID)
  id: string;

  @Field(() => String)
  first_name: string;

  @Field(() => String)
  last_name: string;

  @IsEmail()
  @Field(() => String)
  email: string;

  @Field(() => String)
  role: roles;

  @Field(() => Date)
  created_at: Date;

  @Field(() => Date)
  login_at?: Date;

  @Field(() => Date)
  password_last_updated?: Date;

  @Field(() => [LikesOfProducts])
  likes_products?: LikesOfProducts[];

  password_reset_requests?: number;
  password_reset_requests_timestamps?: Date[];
  failed_login_attempts_timestamps?: Date[];
  password?: string;
  password_reset_token?: string;
  session_token?: string;
}
