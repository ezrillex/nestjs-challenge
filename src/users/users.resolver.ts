import {
  Args,
  ID,
  Parent,
  Query,
  ResolveField,
  Resolver,
} from '@nestjs/graphql';
import { Users } from './users.model';
import { PrismaService } from '../prisma/prisma.service';

@Resolver(() => Users)
export class UsersResolver {
  constructor(private readonly prisma: PrismaService) {}

  @Query(() => Users)
  async user(@Args('id', { type: () => ID }) id: string) {
    return this.prisma.users.findUnique({
      where: {
        id: id,
      },
    });
  }

  @ResolveField()
  async likes_products(@Parent() users: Users) {
    const { id } = users;
    return this.prisma.likesOfProducts.findMany({
      where: {
        user_id: id,
      },
      include: {
        liked_by: {
          select: { id: true },
        },
        likes_product_variation: {
          select: { id: true },
        },
      },
    });
  }
}
