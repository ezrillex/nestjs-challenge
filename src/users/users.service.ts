import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { Users } from './users.model';

@Injectable()
export class UsersService {
  constructor(private readonly prisma: PrismaService) {}

  async findOneByEmail(email: string): Promise<Users> {
    const user = await this.prisma.users.findUnique({
      where: { email: email },
    });
    if (user) {
      return user;
    } else {
      throw new HttpException(
        "Couldn't find your account. Make sure this is the right email.",
        HttpStatus.NOT_FOUND,
      );
    }
  }

  async findOneByID(id: string): Promise<Users> {
    const user = await this.prisma.users.findUnique({
      where: { id: id },
    });
    if (user) {
      return user;
    } else {
      throw new HttpException(
        "Couldn't find an account with the associated id .",
        HttpStatus.NOT_FOUND,
      );
    }
  }

  async ResolveUsersOnCartItemsField(id: string): Promise<Users> {
    const { cart_owner } = await this.prisma.cartItems.findUnique({
      where: {
        id: id,
      },
      select: {
        cart_owner: true,
      },
    });
    return cart_owner;
  }

  async ResolveUsersOnOrdersField(id: string): Promise<Users> {
    const { user } = await this.prisma.orders.findUnique({
      where: {
        id: id,
      },
      select: {
        user: true,
      },
    });
    return user;
  }
}
