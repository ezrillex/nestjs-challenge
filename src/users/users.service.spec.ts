import { Test, TestingModule } from '@nestjs/testing';
import { UsersService } from './users.service';
import { Users } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';

describe('UsersService', () => {
  let service: UsersService;
  let prismaService: PrismaService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [UsersService, PrismaService],
    }).compile();

    service = module.get<UsersService>(UsersService);
    prismaService = module.get(PrismaService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('find one by email user', () => {
    it('should error if email was not found', async () => {
      jest.spyOn(prismaService.users, 'findUnique').mockResolvedValue(null);

      await expect(
        service.findOneByEmail('some@email.com'),
      ).rejects.toThrowErrorMatchingSnapshot('user with email not found');
    });

    it('should return a user if found', async () => {
      const user = { id: 'some id of a user we fetched' } as Users;
      jest.spyOn(prismaService.users, 'findUnique').mockResolvedValue(user);

      await expect(service.findOneByEmail('some@email.com')).resolves.toEqual(
        user,
      );
    });
  });

  describe('find one by user ID', () => {
    it('should throw if id was not found', async () => {
      jest.spyOn(prismaService.users, 'findUnique').mockResolvedValue(null);

      await expect(
        service.findOneByID('some id'),
      ).rejects.toThrowErrorMatchingSnapshot('user with ID not found');
    });

    it('should return user if found', async () => {
      jest.spyOn(prismaService.users, 'findUnique').mockResolvedValue({
        id: '86d739f2-ed29-4a29-8034-e4e2c562e380',
      } as Users);

      await expect(
        service.findOneByID('86d739f2-ed29-4a29-8034-e4e2c562e380'),
      ).resolves.toEqual({
        id: '86d739f2-ed29-4a29-8034-e4e2c562e380',
      });
    });
  });
});
