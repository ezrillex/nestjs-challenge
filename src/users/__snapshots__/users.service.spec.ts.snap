// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersService find one by email user should error if email was not found: user with email not found 1`] = `"Couldn't find your account. Make sure this is the right email."`;

exports[`UsersService find one by user ID should throw if id was not found: user with ID not found 1`] = `"Couldn't find an account with the associated id ."`;
